docs/src/developers/style.md

```diff
@@ -51,6 +51,220 @@ format("test")
 The following sections outline extra style guide points that are not fixed
 automatically by JuliaFormatter.
 
+### Abstract types and composition
+
+Specifying types for method arguments and struct fields is mostly optional in
+Julia. The benefit of abstract method arguments is that it enables functions
+and types from one package to be used with functions and types from another
+package via multiple dispatch.
+
+However, abstractly typed methods have two main drawbacks:
+
+ 1. It's possible to find out that you are working with unexpected types deep
+    in the call chain, potentially leading to hard-to-diagnose [`MethodError`s](https://docs.julialang.org/en/v1/manual/methods/#Defining-Methods).
+ 2. Untyped function arguments can lead to correctness problems if the user's
+    choice of input type does not satisfy the assumptions made by the author of
+    the function.
+
+As a motivating example, consider the following function:
+```jldoctest my_sum
+julia> function my_sum(x)
+           y = 0.0
+           for i in 1:length(x)
+               y += x[i]
+           end
+           return y
+       end
+my_sum (generic function with 1 method)
+```
+This function contains a number of implicit assumptions about the type of `x`:
+ * `x` supports 1-based `getindex` and implements `length`
+ * The element type of `x` supports addition with `0.0`, and then with the
+   result of `x + 0.0`.
+
+!!! info
+    As a motivating example for the second point, [`VariableRef`](@ref) plus
+    `Float64` produces an [`AffExpr`](@ref). Do not assume that `+(::A, ::B)`
+    produces the type `A` or `B`.
+
+`my_sum` works as expected if the user passes in `Vector{Float64}`:
+```jldoctest my_sum
+julia> my_sum([1.0, 2.0, 3.0])
+6.0
+```
+but it doesn't respect input types, for example returning a `Float64` if the
+user passes `Vector{Int}`:
+```jldoctest my_sum
+julia> my_sum([1, 2, 3])
+6.0
+```
+but it throws a `MethodError` if the user passes `String`:
+```jldoctest my_sum
+julia> my_sum("abc")
+ERROR: MethodError: no method matching +(::Float64, ::Char)
+[...]
+```
+This particular `MethodError` is hard to debug, particularly for new users,
+because it mentions `+`, `Float64`, and `Char`, none of which were called or
+passed by the user.
+
+#### Dealing with `MethodError`s
+
+All code must follow the `MethodError` principle:
```

Comment:
All code seems a bit extreme. I can understand providing an opinion and the advantage it can provide in some cases, but overgeneralizing this can lead to a lot of code to produce errors, and may require overly specific methods to avoid ambiguities with the error fallback.

@oxinabox detailed it more here: https://www.oxinabox.net/2020/04/19/Julia-Antipatterns.html#notimplemented-exceptions

Reply:
The experience is that beginner Julia users find MethodErrors hard to deal with. They might not even have a good concept of what a type is, let alone specifics of why `Vector{Float64} <: Vector{Real}` is `false`.

Is "All _public_ code" better? Admittedly, we don't do this perfectly, so it's a standard to strive for, not a fixed rule.

Reply:
I'm not sure I agree with distinguishing between public and private code. Private code should still follow the style guide. Instead of "All code must [...]", what about "Code should [...]"? The style guide is a set of recommendations, so "must" is out of character.

Reply:
I didn't have any comment on this when tagged because, I think JuMP has its own philosphy that is strong and well-rounded and targetting a different audience to my blog post.

But actually now I do have an opinion because JuMP's current doing of this is annoying to me.
because I was like `shadow_price(my_constraint)` just  now
and I got 
```julia
julia> shadow_price.(cons)
ERROR: The shadow price is not defined or not implemented for this type of constraint.
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] shadow_price(con_ref::ConstraintRef{Model, MathOptInterface.ConstraintIndex{MathOptInterface.VectorAffineFunction{Float64}, MathOptInterface.NormOneCone}, VectorShape})
 ```
 
 and my initial instinct was: _"Ok well lets see what types it is defined for"_.
and so I used `methods(shadow_price)`.
```julia
julia> methods(shadow_price)
# 4 methods for generic function "shadow_price":
[1] shadow_price(con_ref::ConstraintRef{<:AbstractModel, MathOptInterface.ConstraintIndex{F, S}}) where {S<:MathOptInterface.LessThan, F} in JuMP at /home/oxinabox/.julia/packages/JuMP/Psd1J/src/constraints.jl:1056
[2] shadow_price(con_ref::ConstraintRef{<:AbstractModel, MathOptInterface.ConstraintIndex{F, S}}) where {S<:MathOptInterface.GreaterThan, F} in JuMP at /home/oxinabox/.julia/packages/JuMP/Psd1J/src/constraints.jl:1069
[3] shadow_price(con_ref::ConstraintRef{<:AbstractModel, MathOptInterface.ConstraintIndex{F, S}}) where {S<:MathOptInterface.EqualTo, F} in JuMP at /home/oxinabox/.julia/packages/JuMP/Psd1J/src/constraints.jl:1082
[4] shadow_price(con_ref::ConstraintRef{<:AbstractModel, <:MathOptInterface.ConstraintIndex}) in JuMP at /home/oxinabox/.julia/packages/JuMP/Psd1J/src/constraints.jl:1017
```
Which of-course lists there as being a method for this.
The one that manually threw the error.

So while i do think JuMP is probably right in having a strong philosphy around these things, 
I will just say it is not without cost in terms of usability to experienced julia programmers.
It is probably worth it to be more friendly to less experienced programmers.
That is a call that should be conciously made however.

Reply:
> I will just say it is not without cost in terms of usability to experienced julia programmers.
> It is probably worth it to be more friendly to less experienced programmers.

Yes, I think we went for the latter.